### PR TITLE
Package initiation and init a Tez module

### PR DESCRIPTION
**Pull request by xvw**
This is a join-work with @mspwn. 
- [x] Initialize the project (and fit with `Nightmare` organization)
- [x] Introduce a `Tez` module
- [x] Add tests for the corner cases of the `Tez` module.

### CODE REVIEW

File: lib/common/tez.ml [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+let to_mutez x = Micro.truncate x
+let from_mutez x = Micro.from_int64 x
+
+module Infix = struct
+  let ( = ) x y = Int64.equal x y
+  let ( <> ) x y = not (x = y)
+  let ( > ) x y = Int64.compare x y > 0
+  let ( >= ) x y = Int64.compare x y >= 0
+  let ( < ) x y = Int64.compare x y < 0
+  let ( <= ) x y = Int64.compare x y <= 0
```
Comment (xhtmlboi) [inline lib/common/tez.ml:119]: I think this part should be tested, you can quickly make mistakes when manipulating the `compare` function.

File: lib/common/tez.ml [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    Format.fprintf ppf "%Li.%04Li" left right
+  | `Five ->
+    let divider = Int64.div Micro.one 10L in
+    let right = Int64.div (Int64.rem value one) divider in
+    Format.fprintf ppf "%Li.%05Li" left right
+  | `Six ->
+    let divider = Micro.one in
+    let right = Int64.div (Int64.rem value one) divider in
+    Format.fprintf ppf "%Li.%06Li" left right
+;;
```
Comment (xhtmlboi) [inline lib/common/tez.ml:168]: As the library embeds the `Nanotez`, it should go up to 9  ;)

File: lib/common/tez.ml [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  let part =
+    match String.length decimal with
+    | 0 -> Some 0L
+    | 1 -> Some 100_000L
+    | 2 -> Some 10_000L
+    | 3 -> Some 1000L
+    | 4 -> Some 100L
+    | 5 -> Some 10L
+    | 6 -> Some 1L
+    | _ -> None
```
Comment (xhtmlboi) [inline lib/common/tez.ml:180]: Same here about the support of `Nanotez`.

File: lib/common/tez.ml [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+   | [ integral; decimal ] ->
+     Option.fold
+       ~none:(Error (`Tez_invalid_string_representation str))
+       ~some:(fun integral ->
+         let integral = from_int64 integral
+         and decimal = decimal_part_from_string str decimal in
+         Infix.(integral |+| decimal))
+       (Int64.of_string_opt integral)
+   | _ -> Error (`Tez_invalid_string_representation str))
+  |> Result.map_error (fun _ -> `Tez_invalid_string_representation str)
```
Comment (xhtmlboi) [inline lib/common/tez.ml:203]: Why this "error nromalization"?

**Review by xhtmlboi — CHANGES_REQUESTED** [review]
Very good start! 
I am still wary of including nanotez in the overall representation of tez. I understand that this is to easily describe the fees for bytes and for gas units but:

- while The Protocol is used to describe the maximum amount of mutez $(2^{63})-1$, by taking on board the nanotez we limit this number to $((2^{63})-1)/1000$
- the rule of rounding is trivial. We go down to the lowest floor.

So I'm not opposed to managing them in this way, but I'm not at all convinced that it's really useful. (But feel free to resolve that part if you did not agree)

And, since `data-encoding` is part of the dependencies set, I suggest to add an `encoding`.

File: lib/common/tez.ml (hunk 4 above)
Comment (xvw) [reply lib/common/tez.ml:203]: In order to reject this kind of string : `123.-451` with the error `Tez_invalid_string_representation "123.-451"` and not with `Tez_negative_amount -451L` which is more clear (imho).

File: lib/common/tez.ml (hunk 1 above)
Comment (xvw) [reply lib/common/tez.ml:119]: Done

**Review by xhtmlboi — APPROVED** [review]
Thank you. I think it's better (and we can think about nanotez when we implement the operation estimates). I'll leave it to @gr-im to merge if he agrees and add that I appreciate these tiny, easy-to-read PRs!

**Review by gr-im — APPROVED** [review]
